ci/env/install-dependencies.sh

```diff
@@ -297,153 +297,160 @@ download_mnist() {
-    fi
-
-    # Try n times; we often encounter OpenSSL.SSL.WantReadError (or others)
-    # that break the entire CI job: Simply retry installation in this case
-    # after n seconds.
-    local status="0";
-    local errmsg="";
-    for _ in {1..3}; do
-      errmsg=$(CC=gcc pip install -Ur "${WORKSPACE_DIR}"/python/requirements.txt 2>&1) && break;
-      status=$errmsg && echo "'pip install ...' failed, will retry after n seconds!" && sleep 30;
-    done
-    if [ "$status" != "0" ]; then
-      echo "${status}" && return 1
-    fi
+  # Array to hold all requirements files to install later
+  requirements_files=()
+  # Single packages to install in sync with files
+  requirements_packages=()
+  # Packages to install _after_ previous files have been installed
+  # (e.g. to install a custom pyarrow or torch version). This
+  # would otherwise conflict with pinned dependencies in our requirements
+  # files.
+  delayed_packages=()
 
-    # Repeat for requirements_test.txt
-    local status="0";
-    local errmsg="";
-    for _ in {1..3}; do
-      errmsg=$(CC=gcc pip install -U -c "${WORKSPACE_DIR}"/python/requirements.txt -r "${WORKSPACE_DIR}"/python/requirements_test.txt 2>&1) && break;
-      status=$errmsg && echo "'pip install ...' failed, will retry after n seconds!" && sleep 30;
-    done
-    if [ "$status" != "0" ]; then
-      echo "${status}" && return 1
-    fi
-
-  fi
+  requirements_files+=("${WORKSPACE_DIR}/python/requirements_test.txt")
 
   if [ "${LINT-}" = 1 ]; then
     install_linters
-    # readthedocs has an antiquated build env.
-    # This is a best effort to reproduce it locally to avoid doc build failures and hidden errors.
-    local python_version
-    python_version="$(python -s -c "import sys; print('%s.%s' % sys.version_info[:2])")"
-    if [ "${OSTYPE}" = msys ] && [ "${python_version}" = "3.8" ]; then
-      { echo "WARNING: Pillow binaries not available on Windows; cannot build docs"; } 2> /dev/null
-    else
-      pip install --use-deprecated=legacy-resolver -r "${WORKSPACE_DIR}"/doc/requirements-doc.txt
-    fi
+
+    requirements_files+=("${WORKSPACE_DIR}/doc/requirements-doc.txt")
   fi
 
   # Additional default doc testing dependencies.
   if [ "${DOC_TESTING-}" = 1 ]; then
-    # For Ray Core and Ray Serve DAG visualization docs test
-    sudo apt-get install -y graphviz
-    pip install -U pydot  # For DAG visualization
-    # For the dataset examples
-    sudo apt-get install -y tesseract-ocr
-    pip install -U pytesseract "spacy>=3" spacy_langdetect
-    python -m spacy download en_core_web_sm
+    # For Ray Core and Ray Serve DAG visualization docs test + dataset examples
+    sudo apt-get install -y graphviz tesseract-ocr
+
+    # For DAG visualization
+    requirements_packages+=("pydot")
+    requirements_packages+=("pytesseract")
+    requirements_packages+=("spacy>=3")
+    requirements_packages+=("spacy_langdetect")
```

Comment:
can we move these into the appropriate requirements files?

Reply:
There is no requirements file for DOC_TESTING, yet. We're also later interacting with spacy. Since this is just using existing logic, can we keep this as is for now?